<?xml version="1.0"?>
<robot name="xArm" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <link name="base_link">
    <visual>
      <origin xyz="0 0 0.02"/>
      <geometry>
        <box size="0.15 0.15 0.04"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
      </material>
    </visual>
  </link>
  
  <joint name="base_joint_5" type="revolute">
    <axis xyz="0 0 1"/>
    <origin xyz="0 0 0.045"/>
    <limit lower="${-90*pi/180}" upper="${150*pi/180}" effort="1" velocity="0.0"/>
    <parent link="base_link"/>
    <child link="upper_base_link"/>
  </joint>

  <link name="upper_base_link">
    <visual>
      <geometry>
        <cylinder radius="0.05" length="0.01"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
      </material>
    </visual>
    <visual>
      <origin xyz="0 0.01 0.01"/>
      <geometry>
        <box size="0.02 0.04 0.02"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
     </material>
    </visual>
  </link>

  <joint name="shoulder_joint_4" type="revolute">
    <axis xyz="1 0 0"/>
    <origin xyz="0 0 0.0125"/>
    <limit lower="${-90*pi/180}" upper="${90*pi/180}" effort="1" velocity="0.0"/>
    <parent link="upper_base_link"/>
    <child link="shoulder_to_elbow_link"/>
  </joint>

  <link name="shoulder_to_elbow_link">
    <visual>
      <origin xyz="0 0 0.05"/>
      <geometry>
        <cylinder radius="0.01" length="0.10"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
     </material>
    </visual>
  </link>

  <joint name="elbow_joint_3" type="revolute">
    <axis xyz="1 0 0"/>
    <origin xyz="0 0 0.10"/>
    <limit lower="${-110*pi/180}" upper="${45*pi/180}" effort="1" velocity="0.0"/>
    <parent link="shoulder_to_elbow_link"/>
    <child link="elbow_to_wrist_link"/>
  </joint>

  <link name="elbow_to_wrist_link">
    <visual>
      <origin xyz="0 0 0.01"/>
      <geometry>
        <box size="0.02 0.02 0.04"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
     </material>
    </visual>
    <visual>
      <origin xyz="0 0 0.05"/>
      <geometry>
        <cylinder radius="0.01" length="0.04"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
     </material>
    </visual>
    <visual>
      <origin xyz="0 0 0.09"/>
      <geometry>
        <box size="0.02 0.02 0.04"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
     </material>
    </visual>
  </link>

  <joint name="wrist_joint_2" type="revolute">
    <axis xyz="1 0 0"/>
    <origin xyz="0 0 0.10"/>
    <limit lower="${-130*pi/180}" upper="${40*pi/180}" effort="1" velocity="0.0"/>
    <parent link="elbow_to_wrist_link"/>
    <child link="inter-wrist_link"/>
  </joint>

  <link name="inter-wrist_link">
    <visual>
      <origin xyz="0 0 0.02"/>
      <geometry>
        <cylinder radius="0.01" length="0.04" />
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
     </material>
    </visual>
    <visual>
      <origin xyz="0 -0.01 0.04"/>
      <geometry>
        <box size="0.02 0.04 0.02"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
     </material>
    </visual>
  </link> 

  <joint name="wrist_joint_1" type="revolute">
    <axis xyz="0 0 1"/>
    <origin xyz="0 0 0.055"/>
    <limit lower="${-90*pi/180}" upper="${90*pi/180}" effort="1" velocity="0.0"/>
    <parent link="inter-wrist_link"/>
    <child link="base_gripper_link"/>
  </joint>

  <link name="base_gripper_link">
    <visual>
      <origin xyz="0 0 0.005"/>
      <geometry>
        <cylinder length= "0.02" radius="0.005"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
     </material>
    </visual>
    <visual>
      <origin xyz="0 -0.01 0.02"/>
      <geometry>
        <box size="0.04 0.02 0.02"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
     </material>
    </visual>
  </link>

 <joint name="right_gripper_joint_0" type="revolute">
    <parent link="base_gripper_link"/>
    <child link="right_gripper_link"/>
    <axis xyz="0 1 0"/>
    <origin xyz="0 0 0.02" rpy="0 0 0"/>
    <limit lower="${-10*pi/180}" upper="${-90*pi/180}" effort="1" velocity="0.0"/>
    <mimic joint="left_gripper_joint_0" multiplier="-1" offset="0" />
  </joint>

  <joint name="left_gripper_joint_0" type="revolute">
    <parent link="base_gripper_link"/>
    <child link="left_gripper_link"/>
    <axis xyz="0 1 0"/>
    <origin xyz="0 0 0.02" rpy="0 0 0"/>
    <limit lower="${-10*pi/180}" upper="${90*pi/180}" effort="1" velocity="0.0"/>
  </joint>

  <link name="right_gripper_link">
    <visual>
      <origin xyz="0.03 0 0" rpy="0 ${90*pi/180} 0"/>
      <geometry>
        <box size="0.005 0.0025 0.06"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
     </material>
    </visual>
  </link>

  <link name="left_gripper_link">
    <visual>
      <origin xyz="-0.03 0 0" rpy="0 ${-90*pi/180} 0"/>
      <geometry>
        <box size="0.005 0.0025 0.06"/>
      </geometry>
      <material name="blue">
       <color rgba="0.2785 0.6 1.0 1"/>
     </material>
    </visual>
  </link>

</robot>